### PR TITLE
MRG: Anonymize cmd

### PR DESCRIPTION
**Pull request by dokato**
Related to issue #4982 .

@agramfort I wonder if we should overwrite a file by default or rather implement some warning with y/n question for the user (+ extra option `-f` force to avoid that).

### CODE REVIEW

File: mne/commands/mne_anonymize.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+def mne_anonymize(fif_fname):
+    """ Call *anonymize_info* on fif file and save.
+    Parameters
+    ----------
+    fif_fname : str
+        Raw fif File
+    """
+    raw = mne.io.read_raw_fif(fif_fname, preload=True)
+    mne.io.anonymize_info(raw.info)
+    raw.save(fif_fname, overwrite=True)
```
Comment (agramfort) [inline mne/commands/mne_anonymize.py]: you should only overwrite if required.

File: mne/commands/mne_anonymize.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    from mne.commands.utils import get_optparser
+
+    parser = get_optparser(__file__)
+
+    options, args = parser.parse_args()
+    if len(args) != 1:
+        parser.print_help()
+        sys.exit(1)
+
+    fname = args[0]
```
Comment (agramfort) [inline mne/commands/mne_anonymize.py]: please use the --file option.

Comment (dokato) [reply mne/commands/mne_anonymize.py]: @agramfort actually here the use of argument was intended, as actually file name is something what we require from user to pass. And from `optparse` documentation:
"note that the phrase “required option” is self-contradictory in English." https://docs.python.org/2/library/optparse.html which kind of makes sense imho

File: mne/commands/mne_anonymize.py (hunk 1 above)
Comment (dokato) [reply mne/commands/mne_anonymize.py]: should we add some prefix `anon-` before anonymised file by default then?

**Comment by agramfort** [comment]
I think we should stick to the historical API which asks that we use the
--file option.

mne anonymize --file foobar_raw.fif

now it's the question of --overwrite and the use of prefix if overwrite is
False

I don't think we should overwrite by default also.

File: mne/commands/mne_anonymize.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import sys
+import mne
+import os
+
+ANONYMIZE_FILE_PREFIX = 'anon'
+
+
+def mne_anonymize(fif_fname, out_fname, dir_name="", overwrite=False):
+    """ Call *anonymize_info* on fif file and save.
```
Comment (agramfort) [inline mne/commands/mne_anonymize.py]: insert empty line here

File: mne/commands/mne_anonymize.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    Parameters
+    ----------
+    fif_fname : str
+        Raw fif File
+    out_fname : str
+        Output file name
+    dir_name : str
+        Path to directory with files
+    overwrite : bool
+        overwriting flag
```
Comment (agramfort) [inline mne/commands/mne_anonymize.py]: please be clearer

File: mne/commands/mne_anonymize.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+def mne_anonymize(fif_fname, out_fname, dir_name="", overwrite=False):
+    """ Call *anonymize_info* on fif file and save.
+    Parameters
+    ----------
+    fif_fname : str
+        Raw fif File
+    out_fname : str
+        Output file name
+    dir_name : str
+        Path to directory with files
```
Comment (agramfort) [inline mne/commands/mne_anonymize.py]: I would just rely on fif_fname. You can split the fif_fname in basedir and file using os module.

File: mne/commands/mne_anonymize.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    if options.file is None:
+        parser.print_help()
+        sys.exit(1)
+
+    fname = options.file
+    out_fname = options.output
+    overwrite = options.overwrite
+
+    dir_name = os.path.split(fname)[0]
+    fname = os.path.basename(fname)
```
Comment (agramfort) [inline mne/commands/mne_anonymize.py]: ok just saw this. I would do this in the function then.

**Comment by dokato** [comment]
Thanks, fixed

**Comment by dokato** [comment]
BTW I know that it might be a subject for another PR, but since you asked me about changing input fif from argument to option, I noticed some inconsistency in command tools:
```
$mne show_info --help
Usage: mne show_info <file>

Show measurement info from .fif file.
```

```
$mne browse_raw --help
Usage: mne browse_raw [options]

Browse raw data.

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  --raw=FILE            Input raw FIF file
```

```
$mne anonymize --help
Usage: mne anonymize [options]

Anonymize .fif file.

Options:
  --version             show program's version number and exit
  -h, --help            show this help message and exit
  -f FILE, --file=FILE  the file to modify.
```
Each command takes input fif in a different way...

**Comment by agramfort** [comment]
you can anonymize more than raw files technically. Any data container with info could be anonymized.

**Comment by dokato** [comment]
Well, but it doesn't explain why in `$mne show_info` it goes with argument and not with `-f` option. But anyway, that's just nitpicking remark.

**Comment by agramfort** [comment]
I guess I wanted to stick to old API so little code would need to be
changed to use mne-python.

**Comment by dokato** [comment]
Okay, got it.

**Comment by dokato** [comment]
I think it's ready now.

File: mne/commands/mne_anonymize.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python
+"""Anonymize .fif file.
+
+You can do for example:
+
+$ mne anonymize sample_audvis_raw.fif
```
Comment (agramfort) [inline mne/commands/mne_anonymize.py]: I don't think this is correct

**Comment by agramfort** [comment]
I suggest to put this on hold as the more I think about this, the more I am convinced that we should do this operation inplace on the file without any copy. We should also be able to anonymize any .fif file that contains a measurement info.

problem is that it requires a deep understanding of the reading of fif tags to be able to do this inplace.

**Comment by larsoner** [comment]
> problem is that it requires a deep understanding of the reading of fif tags to be able to do this inplace.

We could write a function `fid.tell()` us where a given tag's data are, then one that zeroes out that tag's values (for string I guess this would be filling it with `\x00`, for any numeric this would be set it to zeroes). But this will require some care as doing it incorrectly could corrupt the file.

And it will presumably only be doable with `.fif` and not `.fif.gz` files.

**Comment by agramfort** [comment]
+1

that's what the mne-c code does.

File: mne/commands/mne_anonymize.py (hunk 7 above)
Comment (dokato) [reply mne/commands/mne_anonymize.py]: True, fixed.

**Comment by larsoner** [comment]
Closing this for now since it seems like we ultimately need a different solution. Thanks for working on this though @dokato

**Pull request closed without merging**